java/arcs/sdk/EdgeHandles.kt

```diff
@@ -0,0 +1,161 @@
+package arcs.sdk
+
+import arcs.core.entity.Handle
+import arcs.core.entity.ReadCollectionHandle
+import arcs.core.entity.ReadSingletonHandle
+import arcs.core.entity.Storable
+import arcs.core.entity.WriteCollectionHandle
+import arcs.core.entity.WriteSingletonHandle
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.withContext
+
+/*
+ * To facilitate safe read/write access to data handles in running arcs, particles can be
+ * annotated with '@edge' in the manifest. This will cause the following set of changes to
+ * the code generation output:
+ *  - The generated class will be concrete and final instead of an abstract base class.
+ *  - The particle lifecycle will be managed internally to this class; clients will not have
+ *    any access to lifecycle events.
+ *  - Handles will be exposed via the Edge*Handle interfaces given below. All edge handle methods
+ *    are asynchronous and may be invoked at any time after the particle has been created.
+ *  - All read operations will return a CompletableDeferred around the requested value.
+ *  - All write operations will be applied immediately, but the APIs still need to be asynchronous
```

Comment:
It is not clear what the returned job is. Is the write already applied, or do we need to join the job? If so, why not join it as part of the methods here?

Basically with these two comments I am wondering if we can simplify
handles.sw.store(EdgeParticleInternal1("abc")).join()
assertThat(handles.sr.fetch().await()).isEqualTo(EdgeParticleInternal1("abc"))

to be just
handles.sw.store(EdgeParticleInternal1("abc"))
assertThat(handles.sr.fetch()).isEqualTo(EdgeParticleInternal1("abc"))
which seems easier for users of this interface and more idiomatic.

Reply:
The returned `Job` is actually from the standard write API and IIUC is provided as a mechanism for knowing when the write op has settled down the storage stack (it's not actually documented...). They don't actually need to be joined, and reads from the same r/w handle directly after a write should work.

I can update the comment with something like this?

Reply:
Yes a comment would be nice (bonus point for a comment on the main Handle interface as well). If they don't need to be joined, can we remove the join from the tests?

Reply:
Will update. For the tests, it depends on whether they need the store to have propagated downstream or not (e.g. if it's juts local processing on the same handle - not needed; if they're relying on updates flowing to other handles - needed).